Update Unit Tests in PolarizationEffiency Cor

yusufjimoh opened the pull request:
### Description of work
- wrote unit tests to check that wildes method perform well when spinstates is added
- wrote unit tests to check that fredrik method throws exception when tested with wildes-only properties spinstates and flippers
- wrote unit tests to check that fredrikze method can take AddSpinStateToLog property
- refactored unit tests for maintainability

Fixes #38287 

### To test:
- review unit tests
- ensure all unit tests pass

This does not require release notes because it involves updating the unit tests

yusufjimoh committed:
Update Unit Tests in PolarizationEffiency Cor
- wrote unit tests to check that wildes method perform well when spinstates is added
- wrote unit tests to check that fredrik method throws exception when tested with wildes-only properties spinstates and flippers
- wrote unit tests to check that fredrikze method can take AddSpinStateToLog property
- refactored unit tests for maintainability

yusufjimoh committed:
updated hardcoded strings to use private variables

rbauststfc commented on Framework/Algorithms/test/PolarizationEfficiencyCorTest.h:
```diff
@@ -486,10 +427,35 @@ class PolarizationEfficiencyCorTest : public CxxTest::TestSuite {
+    }
+  }
+
+  void checkWorkspaceGroupSize(size_t expectedSize) {
```
This method is checking the size of a specific workspace group (i.e. the one called "out"). Would it be worth renaming it so that it's called `checkOutputWorkspaceGroupSize` to make that clearer?

rbauststfc commented on Framework/Algorithms/test/PolarizationEfficiencyCorTest.h:
```diff
@@ -486,10 +427,35 @@ class PolarizationEfficiencyCorTest : public CxxTest::TestSuite {
+  }
+
+  void checkWorkspaceGroupSize(size_t expectedSize) {
+    WorkspaceGroup_sptr out = AnalysisDataService::Instance().retrieveWS<WorkspaceGroup>("out");
```
Could we define "out" as a constant i.e. `const std::string OUTPUT_GRP_NAME{"out"};`. We use the value in a couple of places, so it would be nice to have it defined in just one place if we can.

rbauststfc commented on Framework/Algorithms/test/PolarizationEfficiencyCorTest.h:
```diff
@@ -486,10 +427,35 @@ class PolarizationEfficiencyCorTest : public CxxTest::TestSuite {
     static std::map<std::string, std::vector<std::string>> const labels = {{"Wildes", {"P1", "P2", "F1", "F2"}},
                                                                            {"Fredrikze", {"Pp", "Ap", "Rho", "Alpha"}}};
-    if (kind == "Wildes" || kind == "Fredrikze") {
+    if (kind == WILDES_METHOD || kind == FREDRIKZE_METHOD) {
```
It's really great for maintainability that we're using the constants here. There are quite a few places in these tests where we're using "Wildes" and "Fredrikze", so could we go through and replace all of those that we can with the constants?

rbauststfc reviewed (commented):
The reduction in code duplication is already looking really good. I have a few suggestions for ways that we might be able to continue to build on this. It would also be really good to try and make more use of constants where we can too.

Commit:
refactored unit tests to reduce redudancies

rbauststfc commented on Framework/Algorithms/test/PolarizationEfficiencyCorTest.h:
```diff
@@ -486,10 +409,35 @@ class PolarizationEfficiencyCorTest : public CxxTest::TestSuite {
+  void setAlgorithmProperties(const std::shared_ptr<PolarizationEfficiencyCor> &alg, const std::string &method = "",
+                              const std::string &analysisMethod = "") {
+    alg->setProperty("OutputWorkspace", OUTPUT_GRP_NAME);
+    if (!method.empty()) {
```
Could we move these two if statements into the `createAlgorithm` method (we'll need to add the `method` and `analysisMethod` parameters to `createAlgorithm` in order to to this)? Then we could remove this separate `setAlgorithmProperties` method, which should further simplify the tests as it seems to be needed in most of the tests.

rbauststfc commented on Framework/Algorithms/test/PolarizationEfficiencyCorTest.h:
```diff
@@ -515,7 +463,7 @@ class PolarizationEfficiencyCorTest : public CxxTest::TestSuite {
       alg->setProperty("F1", ws3);
       alg->setProperty("F2", ws4);
-      alg->setPropertyValue("OutputWorkspace", "dummy");
+      alg->setPropertyValue("OutputWorkspace", OUTPUT_GRP_NAME);
```
This comment should really be on line 458 but I can't add it there, sorry. It looks like there's some code duplication for setting up the `JoinISISPolarizationEfficiencies` algorithm. Could we move this into a helper method (i.e. similar to `createAlgorithm` except this one could be called `createJoinEfficienciesAlgorithm`)?

rbauststfc reviewed (commented):
This is looking much more maintainable than it was, thanks very much for the changes so far! There are a couple of further comments I had.

Commit:
Simplify Tests based on code review suggestions
- created efficiency output workspace method to cater for duplication in createEfficiencyWorkspace
- moved functions from setAlgorithmProperties to createAlgorithm method

rbauststfc commented on Framework/Algorithms/test/PolarizationEfficiencyCorTest.h:
```diff
@@ -486,10 +375,53 @@ class PolarizationEfficiencyCorTest : public CxxTest::TestSuite {
+  }
+
+  MatrixWorkspace_sptr createEfficiencyOutputWorkspace(size_t size = 10, double endX = 10) {
+    auto ws1 = createPointWS(size, 0, endX);
```
I think we need to pass through the `kind` parameter from `createEfficiencies` and use that to determine what type of workspaces to create. It looks like originally we would call `createPointWs` for `points` and `point-short`, but if `kind == "histo"` then we should be calling `createHistoWS`.

rbauststfc commented on Framework/Algorithms/test/PolarizationEfficiencyCorTest.h:
```diff
@@ -50,363 +50,253 @@ class PolarizationEfficiencyCorTest : public CxxTest::TestSuite {
+    WorkspaceGroup_sptr out = AnalysisDataService::Instance().retrieveWS<WorkspaceGroup>(OUTPUT_GRP_NAME);
+    checkOutputWorkspaceGroupSize(4);
 
     for (size_t i = 0; i < out->size(); ++i) {
```
In `test_points` and `test_points_short` we're comparing the output and input workspaces using the `CompareWorkspaces` algorithm. As this code for setting up and calling this algorithm is duplicated between the two tests, could we pull this into a helper method?

rbauststfc reviewed (commented):
There are some really significant improvements in this test class now, thanks very much. There are just a couple more things I noticed, but those should be the last bits I think.

yusufjimoh committed:
remove duplication in test_point methods

rbauststfc commented on Framework/Algorithms/test/PolarizationEfficiencyCorTest.h:
```diff
@@ -502,61 +400,16 @@ class PolarizationEfficiencyCorTest : public CxxTest::TestSuite {
-      alg->setPropertyValue("OutputWorkspace", "dummy");
-      alg->execute();
-      MatrixWorkspace_sptr outWS = alg->getProperty("OutputWorkspace");
+      auto outWS = createEfficiencyOutputWorkspace(10, 10, "histo");
```
I think ideally it would be good to pass the `kind` variable into `createEfficiencyOutputWorkspace` so that we don't create a new string. Perhaps we could simplify these two methods to be:

```
  MatrixWorkspace_sptr createEfficiencyOutputWorkspace(std::string const &kind) {
    const auto numWorkspaces = 4;
    const double endX = 10;

    std::vector<MatrixWorkspace_sptr> workspaces;
    if (kind == "histo") {
      for (int i = 0; i < numWorkspaces; ++i) {
        workspaces.push_back(createHistoWS(10, 0, endX));
      }
    } else if (kind == "points") {
      for (int i = 0; i < numWorkspaces; ++i) {
        workspaces.push_back(createPointWS(10, 0, endX));
      }
    } else if (kind == "points-short") {
      for (int i = 0; i < numWorkspaces; ++i) {
        workspaces.push_back(createPointWS(4, 0, endX));
      }
    } else {
      throw std::logic_error("Unknown efficiency test kind");
    }

    auto alg = AlgorithmFactory::Instance().create("JoinISISPolarizationEfficiencies", -1);
    alg->initialize();
    alg->setChild(true);
    alg->setRethrows(true);
    alg->setProperty("P1", workspaces[0]);
    alg->setProperty("P2", workspaces[1]);
    alg->setProperty("F1", workspaces[2]);
    alg->setProperty("F2", workspaces[3]);
    alg->setPropertyValue("OutputWorkspace", OUTPUT_GRP_NAME);
    alg->execute();

    return alg->getProperty("OutputWorkspace");
  }

  MatrixWorkspace_sptr createEfficiencies(std::string const &kind) {
    static std::map<std::string, std::vector<std::string>> const labels = {
        {WILDES_METHOD, {"P1", "P2", "F1", "F2"}}, {FREDRIKZE_METHOD, {"Pp", "Ap", "Rho", "Alpha"}}};
    if (kind == WILDES_METHOD || kind == FREDRIKZE_METHOD) {
      auto inWS = createWorkspaces(1)[0];
      MatrixWorkspace_sptr ws = WorkspaceFactory::Instance().create(inWS, 4);
      ws->getAxis(0)->setUnit("Wavelength");
      auto axis1 = std::make_unique<TextAxis>(4);
      auto axis1Raw = axis1.get();
      ws->replaceAxis(1, std::move(axis1));
      auto const &current_labels = labels.at(kind);
      for (size_t i = 0; i < ws->getNumberHistograms(); ++i) {
        axis1Raw->setLabel(i, current_labels[i]);
      }
      return ws;
    }

    return createEfficiencyOutputWorkspace(kind);
  }
```

yusufjimoh committed:
refactor createEfficiencyOutputWorkspace method

rbauststfc reviewed (approved):
This refactor has significantly reduced the code duplication and should make it much easier to maintain and extend the tests in future. All code review comments have been addressed and all tests are passing. Thanks very much @yusufjimoh for all your hard work on this one.
